src/dmd/parse.d

```diff
@@ -8973,10 +8976,129 @@ final class Parser(AST) : Lexer
         {
-            s.addComment(combineComments(blockComment, token.lineComment, true));
+            s.addComment(lexer.combineComments(blockComment, token.lineComment, true));
             token.lineComment = null;
         }
     }
+
+    private Loc loc()
+    {
+        return lexer.loc;
+    }
+
+    package ref Loc scanloc() pure nothrow @nogc @safe
+    {
+        return lexer.scanloc;
+    }
+
+    package void scanloc(const ref Loc loc) pure nothrow @nogc @safe
+    {
+        lexer.scanloc = loc;
+    }
+
+    private Token* peek(Token* currentToken)
+    {
+        return lexer.peek(currentToken);
+    }
+
+    package ref Token token() pure nothrow @nogc @safe
+    {
+        return lexer.token;
+    }
+
+    TOK nextToken()
+    {
+        return lexer.nextToken();
+    }
+
+    package TOK peekNext()
+    {
+        return lexer.peekNext();
+    }
+
+    private TOK peekNext2()
+    {
+        return lexer.peekNext2();
+    }
+
+    private Token* peekPastParen(Token* token)
+    {
+        return lexer.peekPastParen(token);
+    }
+
+    package bool errors()
+    {
+        return lexer.errors();
+    }
+
+    package Token* allocateToken() pure nothrow @safe
+    {
+        return lexer.allocateToken();
+    }
+
+    package void error(const(char)* format, ...)
+    {
+        va_list args;
+        va_start(args, format);
+        lexer.error(token.loc, format, args);
+        va_end(args);
+    }
+
+    package void error(const ref Loc loc, const(char)* format, ...)
+    {
+        va_list args;
+        va_start(args, format);
+        lexer.error(loc, format, args);
+        va_end(args);
+    }
+
+    private void errorSupplemental(const ref Loc loc, const(char)* format, ...)
+    {
+        va_list args;
+        va_start(args, format);
+        lexer.errorSupplemental(loc, format, args);
+        va_end(args);
+    }
+
+    private void warning(const ref Loc loc, const(char)* format, ...)
+    {
+        va_list args;
+        va_start(args, format);
+        lexer.warning(loc, format, args);
+        va_end(args);
+    }
+
+    private void warningSupplemental(const ref Loc loc, const(char)* format, ...)
+    {
+        va_list args;
+        va_start(args, format);
+        lexer.warningSupplemental(loc, format, args);
+        va_end(args);
+    }
+
+    private void deprecation(const(char)* format, ...)
+    {
+        va_list args;
+        va_start(args, format);
+        lexer.deprecation(token.loc, format, args);
+        va_end(args);
+    }
+
+    private void deprecation(const ref Loc loc, const(char)* format, ...)
+    {
+        va_list args;
+        va_start(args, format);
+        lexer.deprecation(loc, format, args);
+        va_end(args);
+    }
+
+    private void deprecationSupplemental(const ref Loc loc, const(char)* format, ...)
+    {
+        va_list args;
+        va_start(args, format);
+        lexer.deprecationSupplemental(loc, format, args);
+        va_end(args);
+    }
```

Comment:
These 19 methods that now need awkward wrappers are evidence that this is not a good path of action. For some, probably callers could be replaced to spell `parser.lexer.method` instead of `parser.method`. That would run afoul of the law of Demeter (https://en.wikipedia.org/wiki/Law_of_Demeter) though.

Reply:
I added the wrappers to reduce the number of changes required. Walter often complains that PRs are too big. Some of these methods are called in hundreds of places. With the wrappers they can be removed one at the time, each in a separate PR. Without the wrappers, I need to make all changes at once. It's also not just the parser that needs changing. There is code outside of the parser the access these methods as well:

https://github.com/dlang/dmd/blob/782dc94e6c16d8b5b08e88461ae50878e6cad21f/src/dmd/compiler.d#L92

https://github.com/dlang/dmd/blob/782dc94e6c16d8b5b08e88461ae50878e6cad21f/src/dmd/dsymbolsem.d#L1894

Although, one can question if a lexer needs to have a public API of 19 methods and fields.

Usually everything in the DMD code base is public. This might lead to code starting to use methods and fields they actually have no business of using. It leads to leaky abstractions and goes agains Law of Demeter, as you mentioned above.

Reply:
>Usually everything in the DMD code base is public.

Making more stuff private would be a better way of going about things.

Reply:
The stuff is already used outside of the lexer, making it more difficult.